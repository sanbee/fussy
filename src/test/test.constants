test()
{
  auto x,y;
  print "10pm1+10pm1 = ",10pm1+10pm1,"\n";
  print "2*10pm1    = ",2*10pm1,"\n";
  x=10pm1;
  y:=x;
  print "x=",x," y:=x = ",y,"\n";
  print "x+x=",x+x,"\n";
  print "x+y=",x+y,"\n";;
}
f(){return 10pm1;}
g(){return 10pm1;}
test();
print "f()*g() = ",f()*g(),"\n";
