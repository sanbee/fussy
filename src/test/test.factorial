/* -*- C -*- */
/*
  df(x)
  ----   = f(x) [1/x + 1/(x-1) + 1/(x-2) +...+ 1]
  dx
*/
f(x)
{
  if (x==1) {return x;} else {return x*f(--x);}
}

df(x)
{
  if (x==1) {return x;} else {return 1/x+df(--x);};
}

test()
{
  auto x,i;
  for(i=1;i<20;i++) {x=i+1pm1;print x," ",f(x),"\n";};
}

test();
