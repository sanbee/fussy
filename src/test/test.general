/* -*-C++-*- */
/* A list of statements seperated by seperator ';' */

print "Test for stmt_list\n"
print "------------------\n"
a=1;b=2;a=1;b=2;

/****************************************************
 Leave the previous comment in a single line.  That 
 will ensure that single-line comments as well as
 multiple line comments are tested.  This comments 
 itself is a test of lexical analysis of the input 
 which spans more than one line.

 Following is a test for compund statement
*****************************************************/
{a=1;b=2;}

/****************************************************
 A list of simple and compund statements 
******************************************************/
{a};{b};a;b;

/*****************************************************
 A compund statement with embedded NEWLINES 
******************************************************/
print "Test for stmt_list with embedded NEWLINES\n"
print "-----------------------------------------\n"
{

a=1;


b=2;


c=3;


}

/*************************************************** 
   A compound statement made of simple and other
   compund statements seperated either by a NEWLINE
   or a seperator.  
****************************************************/
{
{a=1;}
{b=2;}
{b=2;}
1;2;3;{b=2;}
1;2;3;{b=2;};1;2;3;



{1;2;3};{b=2;};1;2;3;


{1;2;

}


}

/*****************************************************
 Test for assignment as a function argument 
******************************************************/
print "sin(1.5)=",sin(1.5),"\n"
/* print "sin(a=1.5)=",sin(a=1.5)," a=",a,"\n"*/
/*************************************************
   Just a compund statement, but with NEWLINES
   after and before the starting and closing
   compund statement delimiters ('{' and '}')
**************************************************/
{
{a=1;b=2;}
}

/**********************************************************
 A test to test the print command with strings and numbers 
***********************************************************/
print "this is a test to print a number \"one\" with precision %8.3f ",a%8.3f,"\n"


/**************************************************
 Tests for if statement 
***************************************************/
print "Test for various forms of for-statements\n"
print "----------------------------------------\n"
if (a) 
  { print "a=1",a,"\n";} else 
  { print "Not a\n";}

a=1.1
if (a) 
  { if (a>1) {print "a>1",a,"\n";} else {print "a<=1",a,"\n";} } else 
  { print "a=0\n";}

if (a) a; else b=10;
print "b=",b,"\n"
a=0;
if (a) a; else b=10;;print "b=",b,"\n"

/***************************************************
Test for while-statement 
****************************************************/
print "Test for while-statement\n"
print "------------------------\n"
a=10;while(a>0) {
 if (a>5) a;
 a=a-1;
 if (a<5) -a;
}

/***************************************************
 Test for for-statement on a single line 
****************************************************/
print "Test for for-statement\n"
print "----------------------\n"
for(b=0,a=0;a<10;a=a+1,b=a) print "a=",a%f,"\tb=",b%f,"\n";

/***************************************************
 Test for for-statement not on a single line 
****************************************************/
print "Test for for-statement\n"
print "----------------------\n"
for(b=0,a=0;a<10;a=a+1,b=a) 
 print "a=",a%f,"\tb=",b%f,"\n";

/****************************************************
 Changing the FMT of a variable 
*****************************************************/
a=1;
print "Before change of FMT a = ",a,"\n"

a.=%E

print "After change of FMT a = ",a,"\n"
/****************************************************
   Tests for break statement
****************************************************/
b=0;

print "Test for break statement\n";
print "------------------------\n"
while (a)
{
  for(a=0;a<10;a=a+1) 
   {
     if (a>5) break;
   }
  b=b+1;
  print "Value of b = ",b,"\n";
  if (b>=5)    break;
}

print "Value of a = ",a,"\n";

while(a<100)
{
  a.=%E;
 for(a=0%E;a<100;a=a+1) 

  {
   if (a>=50) break; 
  }

 print "a=";
 print a,"\n"; 

 break;
}

f(x)
{
  auto y;
  print "From f: Arg=",x,"\n";
  x.=%E;
  x;
  y=10;
  print "From f: Auto var y=",y,"\n";
  return y;
}

ff(x) {auto t;print "From ff: Arg=",x,"\n";t=(x+10);t.=%f;return t%f;}

p(func f,x)
{
  auto t;
  print "From p: Arg=",x,"\n";
  t=f(x);
  print "From p: Function called as a arg list of print statement. Result = ",f(x),"\n";
  print "From p: Value of function call in an auto. var. = ",t,"\n";
}

pp(proc p, func f,x)
{
  print "From pp: Arg=",x,"\n";
  p(f,x)
}

print "\nTesting f(10): \n";
print "--------------\n"
f(10);
print "\nTesting ff(10): \n";
print "--------------\n"
ff(10);
print "\nTesting p(ff,10): \n";
print "-----------------\n"
p(ff,10);
print "\nTesting pp(p,ff,10): \n";
print "--------------------\n"
pp(p,ff,10);

inf()
{
  print "Test for an infinite for-loop.  Do a Ctrl-C if you want to do something useful!\n";
  auto i;
  for(i=0;i> -1;i=i+1);
}

infw()
{
  print "Test for an infinite while-loop.  Do a Ctrl-C if you want to do something useful!\n";
  auto i,j;
  j=i=1;
  while(i) {f(i);i=i+2;}
}

/*------------------------------------------------------------------------*/
