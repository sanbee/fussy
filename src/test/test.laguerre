/* -*-C-*- */
/******************************************************************
 * Copyright (c) 2000-2018, 2019 S.Bhatnagar
 *
 *  This file is part of fussy.
 *
 *  fussy is a free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  fussy is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with fussy.  If not, see <https://www.gnu.org/licenses/>.
 *

  History:

    First version.

               Sanjay Bhatnagar, Dec 2001
               sanjay@ncra.tifr.res.in

******************************************************************/
/*
    Recursive generating formula for Laguerre polynomial and it's
    derivative.

    Shows that f(n,x).rms == (df(n,x).val*x.rms) (which is
    f'(n,x)*deltaX) 
*/

f(n,x)
{
  if (n<=0) return 1;
  if (n==1) return 1-x;
  return ((2*n-1-x)*f(n-1,x)-(n-1)*f(n-2,x))/n;
}

df(n,x)
{
  return (n*f(n,x)-n*f(n-1,x))/x;
}

test(x)
{
   auto n,i,y;
   for(i=1;i<10;i++) {print "l(",i%-6.2f,",",x%-6.2f,")= ",f(i,x)," ",fabs(df(i,x).val*x.rms),"\n";};
}

test(3pm1);
