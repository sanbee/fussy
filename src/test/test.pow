/* -*-C-*- */
/* 
   Test for three functions
   f(x)=sin(x)^cos(x) 
   f(x,y)=x^y
   f(x,y)=sin(x)^cos(y)
   
   Error propagation in this expression was tested by working out the
   derivative in calc.  The derivative of the derivative (second
   derivative! ;-)) was also computed in calc.
   
   It is shown that the computed error in f(x) and f(x,y) and the
   error on the computed error is same as that worked out in calc.
   
   The first print statement prints the result of the function.  The
   second print statement prints the result of the propagated error.
   Since this is another expression in x (or y or x and y), it has its
   own propagated error.  The third print statement prints the
   propagated error in the expression for the propagated error in the
   original expression.
*/
msg="Error on the above expression should be = ";

test()
{
  auto tx,ty;
  auto dtx,dty,tmp;

  tx=2pm0.2;
  ty=3pm0.3;
  
  /*----------------------------------------------------------------------------
    f(tx)=sin(tx)^cos(tx)
    ----------------------------------------------------------------------------*/
  print "sin(tx)^cos(tx) = ",sin(tx)^cos(tx),"\n";
  tmp= 0.2*(sin(tx)^cos(tx))* (cos(tx)*cos(tx)/sin(tx) - log(sin(tx))*sin(tx));
  print msg,tmp,"\n";
  tmp=0.2*0.2*(cos(tx)*(cos(tx)*((cos(tx) - 1)*(sin(tx)**(cos(tx) - 2)*cos(tx)) - log(sin(tx))*(sin(tx)**(cos(tx) - 1)*sin(tx))) - sin(tx)**(cos(tx) - 1)*sin(tx)) - sin(tx)**(cos(tx) - 1)*(cos(tx)*sin(tx)) - sin(tx)**cos(tx)*cos(tx) - log(sin(tx))*(sin(tx)*(cos(tx)*(sin(tx)**(cos(tx) - 1)*cos(tx)) - log(sin(tx))*(sin(tx)**cos(tx)*sin(tx))) + sin(tx)**cos(tx)*cos(tx)));
  print msg,tmp ,"\n";
  
  /*----------------------------------------------------------------------------
    f(tx,ty)=tx^ty
    ----------------------------------------------------------------------------*/
  print "tx^ty = ",tx^ty,"\n";
  tmp=(tx^ty)*sqrt(((ty/tx)*0.2)^2+(log(tx)*0.3)^2);
  print msg,tmp,"\n";
  tmp=sqrt((0.2*(0.5*(0.04*(ty**2*((2*ty - 2)*tx**(2*ty - 3)))+0.09*(2*(tx**(2*ty)*log(tx)) / tx + 2*(log(tx)**2*(ty*tx**(2*ty - 1)))))/sqrt(0.04*(ty**2*tx**(2*ty - 2)) + 0.09*(log(tx)**2*tx**(2*ty)))))**2 + (0.3*(0.5*(0.04*(2*(tx**(2*ty - 2)*ty) + 2*(ty**2*(log(tx)*tx**(2*ty - 2))))+0.18*(log(tx)**3*tx**(2*ty)))/sqrt(0.04*(ty**2*tx**(2*ty - 2)) + 0.09*(log(tx)**2*tx**(2*ty)))))**2);
  print msg,tmp,"\n";
  
  /*----------------------------------------------------------------------------
    f(tx,ty)=sin(tx)^cos(ty)
    ----------------------------------------------------------------------------*/
  print "sin(tx)**cos(ty) = ",sin(tx)**cos(ty),"\n";
  
  dtx :=cos(ty)*(sin(tx)**(cos(ty) - 1)*cos(tx));
  dty :=-(log(sin(tx))*(sin(tx)**cos(ty)*sin(ty)));

  tmp=sqrt((dtx*0.2)**2 + (dty*0.3)**2);
  print msg,tmp,"\n";
  
  /*
    The non-partial var version

    print msg,sqrt(((0.2*(cos(ty)*(sin(tx)**(cos(ty) - 1)*cos(tx))))**2) + ((0.3*(-(log(sin(tx))*(sin(tx)**cos(ty)*sin(ty)))))**2)),"\n";
  */
  
  dtx := 0.5*(0.04*(cos(ty)**2*(cos(tx)**2*((2*cos(ty) - 2)*(sin(tx)**(2*cos(ty) - 3)*cos(tx))) - 2*(sin(tx)**(2*cos(ty) - 2)*(cos(tx)*sin(tx))))) + 0.09*(2*(sin(tx)**(2*cos(ty))*(sin(ty)**2*(log(sin(tx))*cos(tx)))) / sin(tx) + 2*(log(sin(tx))**2*(sin(ty)**2*(cos(ty)*(sin(tx)**(2*cos(ty) - 1)*cos(tx))))))) / sqrt(0.04*(cos(ty)**2*(sin(tx)**(2*cos(ty) - 2)*cos(tx)**2)) + 0.09*(log(sin(tx))**2*(sin(tx)**(2*cos(ty))*sin(ty)**2)));

  /*  Phew!  Same expression as above but computed entirely in Calc */
 dtx:=0.5*(0.04*(cos(ty)*(cos(tx)*(cos(ty)*(sin(tx)**(cos(ty) - 1)*(cos(tx)*((cos(ty) - 1)*(sin(tx)**(cos(ty) - 2)*cos(tx)))))) + sin(tx)**(cos(ty) - 1)*(cos(tx)*(cos(ty)*(cos(tx)*((cos(ty) - 1)*(sin(tx)**(cos(ty) - 2)*cos(tx))) - sin(tx)**(cos(ty) - 1)*sin(tx))) - cos(ty)*(sin(tx)**(cos(ty) - 1)*(cos(tx)*sin(tx)))))) + 0.09*(sin(tx)**cos(ty)*(sin(ty)*(log(sin(tx))*(sin(tx)**cos(ty)*(sin(ty)*cos(tx))))) / sin(tx) + log(sin(tx))*(sin(ty)*(log(sin(tx))*(sin(tx)**cos(ty)*(sin(ty)*(cos(ty)*(sin(tx)**(cos(ty) - 1)*cos(tx)))))) + sin(tx)**cos(ty)*(sin(ty)*(sin(tx)**cos(ty)*(sin(ty)*cos(tx)) / sin(tx) + log(sin(tx))*(sin(ty)*(cos(ty)*(sin(tx)**(cos(ty) - 1)*cos(tx))))))))) / sqrt(0.04*(cos(ty)*(sin(tx)**(cos(ty) - 1)*(cos(tx)*(cos(ty)*(sin(tx)**(cos(ty) - 1)*cos(tx)))))) + 0.09*(log(sin(tx))*(sin(tx)**cos(ty)*(sin(ty)*(log(sin(tx))*(sin(tx)**cos(ty)*sin(ty)))))));

  dty :=0.5*(0.09*(log(sin(tx))**2*(2*(sin(tx)**(2*cos(ty))*(sin(ty)*cos(ty))) - 2*(sin(ty)**2*(log(sin(tx))*(sin(tx)**(2*cos(ty))*sin(ty)))))) - 0.04*(2*(cos(ty)**2*(cos(tx)**2*(log(sin(tx))*(sin(tx)**(2*cos(ty) - 2)*sin(ty))))) + 2*(sin(tx)**(2*cos(ty) - 2)*(cos(tx)**2*(cos(ty)*sin(ty)))))) / sqrt(0.04*(cos(ty)**2*(sin(tx)**(2*cos(ty) - 2)*cos(tx)**2)) + 0.09*(log(sin(tx))**2*(sin(tx)**(2*cos(ty))*sin(ty)**2)));
	
  tmp=sqrt((dtx*0.2)**2 + (dty*0.3)**2);
  print msg,tmp,"\n";



  /* The non-partial var version.

  
  print msg,sqrt(0.25*(0.04*(((2*cos(ty) - 2)*(cos(tx)**3*sin(tx)**(2*cos(ty) - 3)) - 2*(sin(tx)**(2*cos(ty) - 1)*cos(tx)))*cos(ty)**2) + 0.09*(2*(sin(tx)**(2*cos(ty) - 1)*(sin(ty)**2*(cos(tx)*log(sin(tx))))) + 2*(log(sin(tx))**2*(sin(tx)**(2*cos(ty) - 1)*(sin(ty)**2*(cos(tx)*cos(ty)))))))**2 / (cos(tx)**2*(cos(ty)**2*sin(tx)**(2*cos(ty) - 2)) + 2.25*(log(sin(tx))**2*(sin(tx)**(2*cos(ty))*sin(ty)**2))) + 0.5625*(0.09*((2*(sin(tx)**(2*cos(ty))*(cos(ty)*sin(ty))) - 2*(sin(tx)**(2*cos(ty))*(sin(ty)**3*log(sin(tx)))))*log(sin(tx))**2) - 0.04*(2*(cos(tx)**2*(cos(ty)**2*(sin(tx)**(2*cos(ty) - 2)*(log(sin(tx))*sin(ty))))) + 2*(cos(tx)**2*(sin(tx)**(2*cos(ty) - 2)*(cos(ty)*sin(ty))))))**2 / (cos(tx)**2*(cos(ty)**2*sin(tx)**(2*cos(ty) - 2)) + 2.25*(log(sin(tx))**2*(sin(tx)**(2*cos(ty))*sin(ty)**2)))),"\n";

    print msg,sqrt(((0.2*(0.5*(0.04*(cos(ty)**2*(cos(tx)**2*((2*cos(ty) - 2)*(sin(tx)**(2*cos(ty) - 3)*cos(tx))) - 2*(sin(tx)**(2*cos(ty) - 2)*(cos(tx)*sin(tx))))) + 0.09*(2*(sin(tx)**(2*cos(ty))*(sin(ty)**2*(log(sin(tx))*cos(tx)))) / sin(tx) + 2*(log(sin(tx))**2*(sin(ty)**2*(cos(ty)*(sin(tx)**(2*cos(ty) - 1)*cos(tx))))))) / sqrt(0.04*(cos(ty)**2*(sin(tx)**(2*cos(ty) - 2)*cos(tx)**2)) + 0.09*(log(sin(tx))**2*(sin(tx)**(2*cos(ty))*sin(ty)**2)))))**2) + ((0.3*(0.5*(0.09*(log(sin(tx))**2*(2*(sin(tx)**(2*cos(ty))*(sin(ty)*cos(ty))) - 2*(sin(ty)**2*(log(sin(tx))*(sin(tx)**(2*cos(ty))*sin(ty)))))) - 0.04*(2*(cos(ty)**2*(cos(tx)**2*(log(sin(tx))*(sin(tx)**(2*cos(ty) - 2)*sin(ty))))) + 2*(sin(tx)**(2*cos(ty) - 2)*(cos(tx)**2*(cos(ty)*sin(ty)))))) / sqrt(0.04*(cos(ty)**2*(sin(tx)**(2*cos(ty) - 2)*cos(tx)**2)) + 0.09*(log(sin(tx))**2*(sin(tx)**(2*cos(ty))*sin(ty)**2)))))**2)),"\n";
  */
};
test();

