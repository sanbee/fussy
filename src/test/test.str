x=10.0;
f(x){auto y;y="test = ";print y,x,"\n";}
f(x);

