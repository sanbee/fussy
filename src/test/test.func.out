fb(x) = 	   2.00000 +/-    0.20000
Following should have same value and error: 
-------------------------------------------
fd(x) = 	  -0.75680 +/-    0.52291
fd1(x) = 	  -0.75680 +/-    0.52291
sin(fa(x)) = 	  -0.75680 +/-    0.52291
sin(x^2) = 	  -0.75680 +/-    0.52291
fd2(x) = 	  -0.75680 +/-    0.52291
Following should have same value and error: 
-------------------------------------------
x*fd(x) = 	  -1.51360 +/-    1.19719
x*fd1(x) = 	  -1.51360 +/-    1.19719
x*sin(fa(x)) = 	  -1.51360 +/-    1.19719
x*sin(x*x) = 	  -1.51360 +/-    1.19719
Following should have same value as above but different error: 
--------------------------------------------------------------
y*fd1(x) = 	  -1.51360 +/-    1.05673
y*sin(fa(x)) = 	  -1.51360 +/-    1.05673
y*sin(x^2) = 	  -1.51360 +/-    1.05673
x*fd2(x) = 	  -1.51360 +/-    1.05673
Value of f(11) =  330.00000 +/-   11.00000
